keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1308 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This proposes a component leader election mechanism that is safer for upgrades
+and rollbacks.
+
+This leader election approach continues to use leases, but with two key
+modifications:
+
+- Instead of a race by component instances to claim the lease, component
+  instances declare candidacy for a lease and a election coordinator claims the
+  lease for the best available candidate. This allows the election coordinator
+  to pick a candidate with the lowest version to ensure that skew rules are not
+  violated.
+- The election coordinator can mark a lease as "end of term" to signal to the
+  current leader to stop renewing the lease. This allows the election
+  coordinator to preempt the current leader and replace it with a better one.
+
+## Motivation
+
+The most common upgrade approach used for Kubernetes control plane components is
+a node-by-node approach where all the component of a control plane node are
+terminated together and then restarted at the new version. This process is
+performed node-by-node across a high availability configuration. 
+
+Systems using node-by-node upgrades:
+
+- Cluster API
+- kubeadm
+- KIND
+
+To respect the [Kubernetes skew
+policy](https://kubernetes.io/releases/version-skew-policy/):
+
+- Upgrades should keep controller managers and schedulers at the *old* version
+  until all apiservers are upgraded.
+- Rollbacks should rollback controller managers and schedulers at the *old*
+  version before any apiservers are rolledback.
+
+But a node-by-node upgrade or rollback does not achieve this today.
+
+- For 3 node control plane upgrade, there is about a 25% chance of a new version
+  of the controller running while old versions of the apiserver are active,
+  resulting in a skew violation. (Consider case where the 2nd node upgraded has
+  the lease)
+- For rollback, it is almost a certainty that skew will be violated.
+
+There is also the possiblity that the lease will be lost by a leader during an
+upgrade or rollback resulting in the version of the controller flip-flopping
+between old and new.
+
+### Goals
+
+During HA upgrades/rollbacks/downgrades, 
+
+Leader elected components:
+
+- Change versions at predictable times
+- Do not violate version skew, even during node-by-node rollbacks
+
+The control plane:
+
+- Can safely canary components and nodes at the new version for an extended
+  period of time, or to pause an upgrade at any step during an upgrade. This
+  enhancement, combined with
+  [UVIP](../4020-unknown-version-interoperability-proxy) helps achieve this.
+  
+
+### Non-Goals
+
+- Change the default leader election for components.
+
+## Proposal
+
+- Offer an opt-in leader election mechanism to:
+  - Elect the candidate with the oldest version available.
+  - Provide a way to preempt the current leader.
```

Comment:
```suggestion
  - Provide a way to preempt the current leader on the upcoming expiry of the term.
```

Reply:
this clarification is still pending. I would still read the text in a way that preempting happens immediately, not by term end.

Reply:
oops sorry I missed this, fixed.